airflow/models.py

```diff
@@ -1288,19 +1288,19 @@ def run(
             verbose=True)
 
         if not runnable and not mark_success:
-            if self.state != State.QUEUED:
-                # If a task's dependencies are met but it can't be run yet then queue it
-                # instead
-                self.state = State.QUEUED
-                msg = "Queuing attempt {attempt} of {total}".format(
-                    attempt=self.try_number % (task.retries + 1) + 1,
-                    total=task.retries + 1)
-                logging.info(hr + msg + hr)
-
-                self.queued_dttm = datetime.now()
-                msg = "Queuing into pool {}".format(self.pool)
-                logging.info(msg)
-                session.merge(self)
+            # we might have hit concurrency limits, which means we probably
+            # have been running prematurely.
+            self.state = State.UP_FOR_RETRY
```

Comment:
QUEUED cannot be done, you would end up scheduling tasks twice of even more often (this happened before). NONE and SCHEDULED are scheduler states outgoing tasks (to the executor) all have QUEUED as state. They NONE and SCHEDULED states really should not be set by task instances. That is the reason why we are having this issue.

NONE can be up for debate (as it is a kind of ill defined state), but I also think that hitting a concurrency limit should include a back off period. Setting the state to NONE (or SCHEDULED) for that matter does not do that, and is metaphorically speaking, a scattergun approach: let's hope one will start running.

I understand the issue from the user's perspective and by setting it to something else than UP_FOR_RETRY we hide the issue in the scheduler that we are not checking the concurrency limits. I could clarify this in the logging, so users get educated quickly?